### PR TITLE
 [BEAM-11358] Remove unused endOfStreamMonitor from nexmark and remove redundant distributions from monitor

### PR DESCRIPTION
**Pull request by rHermes**
The `endOfStreamMonitor` in the `NexmarkQuery` class is not used. The monitor class also had start and end distributions, which used the same values. This PR removes the unused monitor and merges the start and end distributions.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [X] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [X] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [X] Update `CHANGES.md` with noteworthy changes.
 - [X] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by rHermes** [comment]
R: @echauchot 
R: @aromanenko-dev

**Comment by rHermes** [comment]
The spotless check failed with a file I have not edited :confused: I can apply the spotless fix and upload that here if you want?

**Comment by rHermes** [comment]
Run Java PreCommit

**Comment by rHermes** [comment]
Seems I've been a bit quick in the trigger here and changed the logic after merging the two distributions. I'll push a fix later this evening, sorry for the slip!

**Comment by rHermes** [comment]
I've fixed the error, but the error now is not in a part of the code that I haven't touched.

**Comment by aromanenko-dev** [comment]
Run Java PreCommit

**Comment by aromanenko-dev** [comment]
@rHermes Thanks! Could you create a Jira for this and squash all your commits into one?

**Comment by rHermes** [comment]
> @rHermes Thanks! Could you create a Jira for this and squash all your commits into one?

Will do it after work today :)

**Comment by rHermes** [comment]
OK, squashed and created a JIRA issue.

**Comment by rHermes** [comment]
@aromanenko-dev does it look good now?

**Comment by aromanenko-dev** [comment]
@rHermes Thanks! Could you elaborate changes regarding this part `This PR ... merges the start and end distributions.` ?

**Comment by aromanenko-dev** [comment]
@echauchot Could you take a look on this PR since you worked on Nexmark too?

**Comment by rHermes** [comment]
> @rHermes Thanks! Could you elaborate the changes regarding this part `This PR ... merges the start and end distributions.` ? Why it was done?

Because they are updated with exactly the same values, they will always be exactly the same, so they could be merged. If you look at the removed lines in Monitor you will see that they where set with exactly the same values

**Comment by echauchot** [comment]
@rHermes thanks for your work ! taking a look

**Comment by rHermes** [comment]
> @rHermes thanks for your work ! taking a look

Any update on this?

**Comment by echauchot** [comment]
@rHermes sorry I'm late on that I started the review and then I was dragged on other priorities. I think I could finish by the end of the week.

File: sdks/java/testing/nexmark/src/main/java/org/apache/beam/sdk/nexmark/Monitor.java [hunk 1]
```diff
@@ -39,20 +39,15 @@
   private class MonitorDoFn extends DoFn<T, T> {
     final Counter elementCounter = Metrics.counter(name, prefix + ".elements");
     final Counter bytesCounter = Metrics.counter(name, prefix + ".bytes");
-    final Distribution startTime = Metrics.distribution(name, prefix + ".startTime");
-    final Distribution endTime = Metrics.distribution(name, prefix + ".endTime");
-    final Distribution startTimestamp = Metrics.distribution(name, prefix + ".startTimestamp");
-    final Distribution endTimestamp = Metrics.distribution(name, prefix + ".endTimestamp");
+    final Distribution systemTimestamp = Metrics.distribution(name, prefix + ".systemTimestamp");
```
Comment (echauchot) [inline sdks/java/testing/nexmark/src/main/java/org/apache/beam/sdk/nexmark/Monitor.java:42]: it is not a timeStamp, it is in processing time (the wall clock) as opposed to a timestamp which is a component of an element put by the source. Please rename to "processingTime"

File: sdks/java/testing/nexmark/src/main/java/org/apache/beam/sdk/nexmark/NexmarkLauncher.java [hunk 2]
```diff
@@ -302,7 +302,7 @@ private NexmarkPerf currentPerf(
       effectiveEnd = eventEnd;
     }
 
-    if (effectiveEnd >= 0 && eventStart >= 0 && effectiveEnd >= eventStart) {
+    if (eventStart >= 0 && effectiveEnd >= eventStart) {
```
Comment (echauchot) [inline sdks/java/testing/nexmark/src/main/java/org/apache/beam/sdk/nexmark/NexmarkLauncher.java:305]: true that this expression is equivalent and simpler but less readable. I would prefer a revert

File: sdks/java/testing/nexmark/src/main/java/org/apache/beam/sdk/nexmark/NexmarkLauncher.java [hunk 3]
```diff
@@ -336,7 +336,7 @@ private NexmarkPerf currentPerf(
       perf.startupDelaySec = (eventStart - startMsSinceEpoch) / 1000.0;
     }
 
-    if (resultStart >= 0 && eventStart >= 0 && resultStart >= eventStart) {
+    if (eventStart >= 0 && resultStart >= eventStart) {
```
Comment (echauchot) [inline sdks/java/testing/nexmark/src/main/java/org/apache/beam/sdk/nexmark/NexmarkLauncher.java:339]: same here

File: sdks/java/testing/nexmark/src/main/java/org/apache/beam/sdk/nexmark/Monitor.java [hunk 4]
```diff
@@ -39,20 +39,15 @@
...
     @ProcessElement
     public void processElement(ProcessContext c) {
       elementCounter.inc();
       bytesCounter.inc(c.element().sizeInBytes());
-      long now = System.currentTimeMillis();
-      startTime.update(now);
-      endTime.update(now);
-      startTimestamp.update(c.timestamp().getMillis());
-      endTimestamp.update(c.timestamp().getMillis());
+      systemTimestamp.update(System.currentTimeMillis());
```
Comment (echauchot) [inline sdks/java/testing/nexmark/src/main/java/org/apache/beam/sdk/nexmark/Monitor.java:49]: agree with your simplification for these reasons:
- for each element processed, the distribution metrics start and end will be updated with the same value and, as the distribution reports min, max and average, and, as in _MetricsReader_ we are requesting the minimum and the maximum, there is no impact of having a single distribution
- the methods in _MetricsReader_ use a SplitIterator on the distributions that is not immutable for the underlying source (which could explain that there was 2 distributions created) but it does only reading (by calling _getMin()_ and _getMax()_) so there will be no side effect on the underlying distribution.

**Comment by echauchot** [comment]
Run Direct Runner Nexmark Tests

**Review by echauchot — CHANGES_REQUESTED** [review]
Thanks for your work @rHermes !
Only some minor fixes and an important comment.
Nexmark execution times are similar to CI. So LGTM
On piece of advice, it is your second cleaning PR on nexmark. Next time, could you please consider submitting cleaning PRs all in a single PR with separate commits with explicit messages? thanks

**Comment by rHermes** [comment]
> Thanks for your work @rHermes !
> Only some minor fixes and an important comment.
> Nexmark execution times are similar to CI. So LGTM
> On piece of advice, it is your second cleaning PR on nexmark. Next time, could you please consider submitting cleaning PRs all in a single PR with separate commits with explicit messages? thanks

I thought it was better practice to create separate PRs, but it's easier for me to just combine them, so I'll do that next time :)

**Comment by echauchot** [comment]
@rHermes yes general good practice is to separate PRs that deal with different subjects. But for cleaning PRs, it is the same subject. There a lot of leftovers in nexmark what I would like to avoid is tens of PRs that remove only a couple of fields.

**Comment by rHermes** [comment]
> @rHermes yes general good practice is to separate PRs that deal with different subjects. But for cleaning PRs, it is the same subject. There a lot of leftovers in nexmark what I would like to avoid is tens of PRs that remove only a couple of fields.

Got it :+1: 

I've addressed your comments from the review, and rebased on top of master.

File: sdks/java/testing/nexmark/src/main/java/org/apache/beam/sdk/nexmark/Monitor.java (hunk 1 above)
Comment (echauchot) [reply sdks/java/testing/nexmark/src/main/java/org/apache/beam/sdk/nexmark/Monitor.java]: please rename also the metric name

**Review by echauchot — CHANGES_REQUESTED** [review]
@rHermes sorry for the delay. I took a llok at your changes. you just forgot a rename and I can merge.

**Comment by rHermes** [comment]
> @rHermes sorry for the delay. I took a llok at your changes. you just forgot a rename and I can merge.

Ahh, my bad. Fixed now :)

**Comment by echauchot** [comment]
Run Direct Runner Nexmark Tests

**Comment by echauchot** [comment]
Run Direct Runner Nexmark Tests

**Review by echauchot — APPROVED** [review]
@rHermes thanks for your work! LGTM. Nexmark last results are correct, merging

**Comment by echauchot** [comment]
my bad, I merged but I forgot to tell you to rename the commit with ticket number and to squash them. It is not that serious, just not very clear in history. you'll know for next time.

**Pull request merged**